src/provider/shplonk.rs

```diff
@@ -0,0 +1,660 @@
+    let u = vec![r, -r, r * r];
+
+    let evals = Self::compute_evals(&polys, &u);
+
+    // Phase 3
+    // Compute B(x) = f_0(x) + q * f_1(x) + ... + q^(k-1) * f_{k-1}(x)
+    let q = HyperKZG::<E, NE>::get_batch_challenge(&evals, transcript);
+    let batched_Pi: UniPoly<E::Fr> = polys.into_iter().map(UniPoly::new).rlc(&q);
+
+    // Q(x), R(x) = P(x) / D(x), where D(x) = (x - r) * (x + r) * (x - r^2) = 1 * x^3 - r^2 * x^2 - r^2 * x + r^4
+    let D = UVKZGPoly::new(vec![u[2] * u[2], -u[2], -u[2], E::Fr::from(1)]);
+    let (Q_x, R_x) = batched_Pi.divide_with_q_and_r(&D).unwrap();
+
+    let C_Q = <NE::CE as CommitmentEngineTrait<NE>>::commit(ck, &Q_x.coeffs)
+      .comm
+      .to_affine();
+
+    let a = Self::compute_a(&C_Q, transcript);
+
+    // K(x) = P(x) - Q(x) * D(a) - R(a), note that R(a) should be subtracted from a free term of polynomial
+    let K_x = Self::compute_k_polynomial(&batched_Pi, &Q_x, &D, &R_x, a);
+
+    // TODO: since this is a usual KZG10 we should use it as utility instead
+    // H(x) = K(x) / (x - a)
+    let divisor = UVKZGPoly::new(vec![-a, E::Fr::from(1)]);
+    let (H_x, _) = K_x.divide_with_q_and_r(&divisor).unwrap();
+
+    let C_H = <NE::CE as CommitmentEngineTrait<NE>>::commit(ck, &H_x.coeffs)
+      .comm
+      .to_affine();
+
+    Ok(EvaluationArgument::<E> {
+      comms,
+      evals,
+      R_x: R_x.coeffs,
+      C_Q,
+      C_H,
+    })
+  }
+
+  /// A method to verify purported evaluations of a batch of polynomials
+  fn verify(
+    vk: &KZGVerifierKey<E>,
+    transcript: &mut <NE as NovaEngine>::TE,
+    _C: &Commitment<NE>,
+    point: &[E::Fr],
+    _P_of_x: &E::Fr,
+    pi: &EvaluationArgument<E>,
+  ) -> Result<(), NovaError> {
+    let r = HyperKZG::<E, NE>::compute_challenge(&pi.comms, transcript);
+    let u = [r, -r, r * r];
+
+    if pi.evals.len() != u.len() {
+      return Err(NovaError::ProofVerifyError);
+    }
+    if pi.R_x.len() != u.len() {
+      return Err(NovaError::ProofVerifyError);
+    }
+
+    // TODO:
+    // insert _P_of_x into every pi.evals_i[last]
+    // insert _C into pi.comms[0]
+    // compute commitment for eval and insert it into pi.comms[last]
+
+    let q = HyperKZG::<E, NE>::get_batch_challenge(&pi.evals, transcript);
+    let q_powers = HyperKZG::<E, NE>::batch_challenge_powers(q, pi.comms.len());
+
+    let R_x = UVKZGPoly::new(pi.R_x.clone());
+
+    let mut evals_at_r = vec![];
+    let mut evals_at_minus_r = vec![];
+    let mut evals_at_r_squared = vec![];
+    for (i, evals_i) in pi.evals.iter().enumerate() {
+      if i == 0 {
+        evals_at_r = evals_i.clone();
+      }
+      if i == 1 {
+        evals_at_minus_r = evals_i.clone();
+      }
+      if i == 2 {
+        evals_at_r_squared = evals_i.clone();
+      }
+
+      let mut batched_eval = E::Fr::ZERO;
+      evals_i
+        .iter()
+        .zip_eq(q_powers.iter())
+        .for_each(|(eval, q_i)| {
+          batched_eval += *eval * q_i;
+        });
```

Comment:
I believe this is `UniPoly::ref_cast(evals_i).evaluate(q)`, eliminating the need for `q_powers`.